gpt4all-kotlin/src/jvmMain/kotlin/com/xebia/functional/gpt4all/GPT4All.kt

```diff
@@ -117,51 +111,52 @@ interface GPT4All : AutoCloseable, Chat, Completion {
        * @return A Flow of ChatCompletionChunk objects representing the generated chat completions.
        */
-      override suspend fun createChatCompletions(request: ChatCompletionRequest): Flow<ChatCompletionChunk> =
-        with(request) {
-          val prompt: String = messages.buildPrompt()
-          val config = LLModel.config()
-            .withTopP(request.topP.toFloat())
-            .withTemp(request.temperature.toFloat())
-            .withRepeatPenalty(request.frequencyPenalty.toFloat())
+      override suspend fun createChatCompletions(request: ChatCompletionRequest): Flow<ChatCompletionChunk> {
+        val prompt: String = request.messages.buildPrompt()
+        val config = with(request) {
+          LLModel.config()
+            .withTopP(topP.toFloat())
+            .withTemp(temperature.toFloat())
+            .withRepeatPenalty(frequencyPenalty.toFloat())
             .build()
+        }
 
-          val originalOut = System.out // Save the original standard output
+        val originalOut = System.out // Save the original standard output
 
-          return coroutineScope {
-            val channel = Channel<String>(capacity = UNLIMITED)
+        val channel = Channel<String>(capacity = UNLIMITED)
 
-            val outputStream = object : OutputStream() {
-              override fun write(b: Int) {
-                val c = b.toChar()
-                channel.trySend(c.toString())
-              }
-            }
+        val outputStream = object : OutputStream() {
+          override fun write(b: Int) {
+            val c = b.toChar()
+            channel.trySend(c.toString())
+          }
+        }
 
-            val printStream = PrintStream(outputStream, true, StandardCharsets.UTF_8)
+        val printStream = PrintStream(outputStream, true, StandardCharsets.UTF_8)
 
-            fun toChunk(text: String?): ChatCompletionChunk =
-              ChatCompletionChunk(
-                UUID.randomUUID().toString(),
-                System.currentTimeMillis().toInt(),
-                path.name,
-                listOf(ChatChunk(delta = ChatDelta(Role.ASSISTANT, text))),
-                Usage.ZERO,
-              )
+        fun toChunk(text: String?): ChatCompletionChunk = ChatCompletionChunk(
+          UUID.randomUUID().toString(),
+          System.currentTimeMillis().toInt(),
+          path.name,
+          listOf(ChatChunk(delta = ChatDelta(Role.ASSISTANT, text))),
+          Usage.ZERO,
+        )
 
-            val flow = channel.consumeAsFlow().map { toChunk(it) }
+        return channel
+          .consumeAsFlow()
+          .map(::toChunk)
+          .onStart {
+            System.setOut(printStream) // Set the standard output to the print stream
 
-            launch(Dispatchers.IO) {
-              System.setOut(printStream) // Set the standard output to the print stream
+            thread(isDaemon = true) { // generate in background and emit values to flow
```

Comment:
Introducing a coroutine scope inside this function was the original issue. The only solution I saw is creating a thread. Any better suggestions?
I tried to minimize devastation by spawning it inside the `onStart` event.

Reply:
> Introducing a coroutine scope inside this function was the original issue.

Can you elaborate? These two should be equivalent, except that the `launch` was probably not properly scoped to the `Flow`. I think this could work.

```kotlin
merge(
  suspend {
    generateCompletion(prompt, config, request.streamToStandardOut)
    channel.close()
  }.asFlow().flowOn(Dispatchers.IO),
  channel
    .consumeAsFlow()
    ...
)
```

`merge` runs two `Flow` in parallel. Here we use a `Flow` that emits no elements to execute the side-effect in parallel.

Reply:
I tried to incorporate your points in the latest commit.

Clarifying: Previously, a coroutineScope was used to `lauch` the generation of next tokens _supposedly_ in the background. As the coroutineScope call didn't finish until the launch job finishes, the function didn't return and thus there was no streaming effect.

You suggested merging two flows and make flow generating text run on IO. Better than creating thread! That appears to be working. Thanks. Instead of converting a suspend function to a flow, I had to create an `emptyFlow()` and call `onStart`.